streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java

```diff
@@ -0,0 +1,468 @@
+    public static class CapturedPunctuator {
+        private final long intervalMs;
+        private final PunctuationType type;
+        private final Punctuator punctuator;
+
+        // mutable:
+        private boolean cancelled = false;
+
+        private CapturedPunctuator(final long intervalMs, final PunctuationType type, final Punctuator punctuator) {
+            this.intervalMs = intervalMs;
+            this.type = type;
+            this.punctuator = punctuator;
+        }
+
+        public long getIntervalMs() {
+            return intervalMs;
+        }
+
+        public PunctuationType getType() {
+            return type;
+        }
+
+        public Punctuator getPunctuator() {
+            return punctuator;
+        }
+
+        public void cancel() {
+            this.cancelled = true;
+        }
+
+        public boolean cancelled() {
+            return cancelled;
+        }
+    }
+
+    private final List<CapturedPunctuator> punctuators = new LinkedList<>();
+
+    private class CapturedForward {
+        /*Nullable*/ private final Integer childIndex;
+        /*Nullable*/ private final String childName;
+        private final KeyValue kv;
+
+        private CapturedForward(final KeyValue kv) {
+            this.childIndex = null;
+            this.childName = null;
+            this.kv = kv;
+        }
+
+        private CapturedForward(final Integer childIndex, final KeyValue kv) {
+            this.childIndex = childIndex;
+            this.childName = null;
+            this.kv = kv;
+        }
+
+        private CapturedForward(final String childName, final KeyValue kv) {
+            this.childIndex = null;
+            this.childName = childName;
+            this.kv = kv;
+        }
+    }
+
+    private List<CapturedForward> capturedForwards = new LinkedList<>();
+
+    private boolean committed = false;
+
+    // contructors ================================================
+
+    /**
+     * Create a {@link MockProcessorContext} with dummy {@code config} and {@code taskId} and {@code null} {@code stateDir}.
+     * Most unit tests using this mock won't need to know the taskId,
+     * and most unit tests should be able to get by with the
+     * {@link InMemoryKeyValueStore}, so the stateDir won't matter.
+     */
+    public MockProcessorContext() {
+        //noinspection DoubleBraceInitialization
+        this(
+            new Properties() {
+                {
+                    put(StreamsConfig.APPLICATION_ID_CONFIG, "");
+                    put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "");
+                }
+            },
+            new TaskId(0, 0),
+            null);
+    }
+
+    /**
+     * Create a {@link MockProcessorContext} with dummy {@code taskId} and {@code null} {@code stateDir}.
+     * Most unit tests using this mock won't need to know the taskId,
+     * and most unit tests should be able to get by with the
+     * {@link InMemoryKeyValueStore}, so the stateDir won't matter.
```

Comment:
I'd suggest when editing on `https://kafka.apache.org/11/documentation/streams/developer-guide/testing.html` state clearly how to switch from an in-memory store to a persistent store when move from unit test code to production code.

Reply:
good idea. Maybe I'll include a link to that page in the javadoc?

Reply:
Sounds good.

Reply:
This link might break in the future. Do you really want to put it in?

Reply:
I think it's nice to link to it from here. We can add the stablest link we're aware of (I'm not sure what `/11/` is for). If it does break, it will at least clue the reader in that there *is* more documentation out there, and they can search until they find it.

I have had this exact thing happen to me before, and found useful information that way.

But I understand it's also setting us up to have a broken link in our docs (since there's no process in place to verify it resolves), which doesn't make us look great if/when it breaks.

I don't feel strongly either way.

Reply:
Agreed. I've removed it.